plugin/evm/vm.go

```diff
@@ -1023,92 +1019,127 @@ func (vm *VM) initBlockBuilding() error {
 		},
 	}
 
-	atomicTxGossipClient, err := vm.router.RegisterAppProtocol(atomicTxGossipProtocol, atomicTxGossipHandler, vm.validators)
+	atomicTxPullGossipClient, err := vm.router.RegisterAppProtocol(atomicTxPullGossipProtocol, atomicTxPullGossipHandler, vm.validators)
 	if err != nil {
 		return err
 	}
 
 	var (
-		ethTxGossiper    gossip.Gossiper
-		atomicTxGossiper gossip.Gossiper
+		ethTxPullGossiper    gossip.Gossiper
+		atomicTxPullGossiper gossip.Gossiper
 	)
-	ethTxGossiper, err = gossip.NewPullGossiper[GossipEthTx, *GossipEthTx](
+
+	ethTxPullGossiper, err = gossip.NewPullGossiper[GossipEthTx, *GossipEthTx](
 		ethTxGossipConfig,
 		vm.ctx.Log,
 		ethTxPool,
-		ethTxGossipClient,
+		ethTxPullGossipClient,
 		vm.sdkMetrics,
 	)
 	if err != nil {
 		return err
 	}
-	ethTxGossiper = gossip.ValidatorGossiper{
-		Gossiper:   ethTxGossiper,
+
+	ethTxPullGossiper = gossip.ValidatorGossiper{
+		Gossiper:   ethTxPullGossiper,
 		NodeID:     vm.ctx.NodeID,
 		Validators: vm.validators,
 	}
 
 	vm.shutdownWg.Add(1)
 	go func() {
-		gossip.Every(ctx, vm.ctx.Log, ethTxGossiper, gossipFrequency)
+		gossip.Every(ctx, vm.ctx.Log, ethTxPullGossiper, gossipFrequency)
 		vm.shutdownWg.Done()
 	}()
 
-	atomicTxGossiper, err = gossip.NewPullGossiper[GossipAtomicTx, *GossipAtomicTx](
+	atomicTxPullGossiper, err = gossip.NewPullGossiper[GossipAtomicTx, *GossipAtomicTx](
 		atomicTxGossipConfig,
 		vm.ctx.Log,
 		vm.mempool,
-		atomicTxGossipClient,
+		atomicTxPullGossipClient,
 		vm.sdkMetrics,
 	)
 	if err != nil {
 		return err
 	}
-	atomicTxGossiper = gossip.ValidatorGossiper{
-		Gossiper:   atomicTxGossiper,
+
+	atomicTxPullGossiper = gossip.ValidatorGossiper{
+		Gossiper:   atomicTxPullGossiper,
 		NodeID:     vm.ctx.NodeID,
 		Validators: vm.validators,
 	}
 
 	vm.shutdownWg.Add(1)
 	go func() {
-		gossip.Every(ctx, vm.ctx.Log, atomicTxGossiper, gossipFrequency)
+		gossip.Every(ctx, vm.ctx.Log, atomicTxPullGossiper, gossipFrequency)
 		vm.shutdownWg.Done()
 	}()
 
+	gossipStats := NewGossipStats()
+	vm.gossipHandler = NewGossipHandler(vm, gossipStats)
+
+	ethTxPushGossipHandler := ethTxPushGossipHandler{
+		NoOpHandler:   p2p.NoOpHandler{},
+		GossipHandler: vm.gossipHandler,
+		codec:         vm.networkCodec,
+	}
+
+	ethTxPushGossipClient, err := vm.router.RegisterAppProtocol(ethTxPushGossipProtocol, ethTxPushGossipHandler, nil)
+	if err != nil {
+		return err
+	}
+
+	atomicTxPushGossipHandler := atomicTxPushGossipHandler{
+		NoOpHandler:   p2p.NoOpHandler{},
+		GossipHandler: vm.gossipHandler,
+		codec:         vm.networkCodec,
+	}
+
+	atomicTxPushGossipClient, err := vm.router.RegisterAppProtocol(atomicTxPushGossipProtocol, atomicTxPushGossipHandler, nil)
+	if err != nil {
+		return err
+	}
+
+	vm.gossiper = vm.createGossiper(gossipStats, ethTxPushGossipClient, atomicTxPushGossipClient)
+	vm.builder = vm.NewBlockBuilder(vm.toEngine)
+	vm.builder.awaitSubmittedTxs()
+	vm.Network.SetGossipHandler(vm.gossipHandler)
```

Comment:
@aaronbuchwald do you think we should first set the handler then call `awaitSubmittedTxs`?
This is unrelated with this PR.